samples/dfu/src/flash.c

```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright (c) 2021 Golioth, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <logging/log.h>
+LOG_MODULE_DECLARE(golioth_dfu);
+
+#include <stdio.h>
+#include <storage/flash_map.h>
+
+#include "flash.h"
+
+/* FLASH_AREA_ID() values used below are auto-generated by DT */
+#ifdef CONFIG_TRUSTED_EXECUTION_NONSECURE
+#define FLASH_AREA_IMAGE_PRIMARY FLASH_AREA_ID(image_0_nonsecure)
+#define FLASH_AREA_IMAGE_SECONDARY FLASH_AREA_ID(image_1_nonsecure)
+#else
+#define FLASH_AREA_IMAGE_PRIMARY FLASH_AREA_ID(image_0)
+#if FLASH_AREA_LABEL_EXISTS(image_1)
+#define FLASH_AREA_IMAGE_SECONDARY FLASH_AREA_ID(image_1)
+#endif
+#endif /* CONFIG_TRUSTED_EXECUTION_NONSECURE */
+
+char current_version_str[sizeof("255.255.65535")];
+
+static int current_version_init(const struct device *dev)
+{
+	struct mcuboot_img_header hdr;
+	int written;
+	int err;
+
+	err = boot_read_bank_header(FLASH_AREA_IMAGE_PRIMARY, &hdr, sizeof(hdr));
+	if (err) {
+		LOG_ERR("Failed to read primary area (%u) header: %d",
+			FLASH_AREA_IMAGE_PRIMARY, err);
+		return err;
+	}
+
+	written = snprintf(current_version_str, sizeof(current_version_str),
+			   "%u.%u.%u",
+			   (unsigned int) hdr.h.v1.sem_ver.major,
+			   (unsigned int) hdr.h.v1.sem_ver.minor,
+			   (unsigned int) hdr.h.v1.sem_ver.revision);
+	if (written >= sizeof(current_version_str)) {
+		LOG_ERR("Version string is too long!");
+		current_version_str[0] = '\0';
+		return -ENOMEM;
+	}
+
+	return 0;
+}
+
+SYS_INIT(current_version_init, APPLICATION, CONFIG_APPLICATION_INIT_PRIORITY);
+
+/**
+ * Determines if the specified area of flash is completely unwritten.
+ *
+ * @note This is a copy of zephyr_img_mgmt_flash_check_empty() from mcumgr.
+ */
+static int flash_area_check_empty(const struct flash_area *fa,
+				  bool *out_empty)
+{
+	uint32_t data[16];
+	off_t addr;
+	off_t end;
+	int bytes_to_read;
+	int rc;
+	int i;
+
+	__ASSERT_NO_MSG(fa->fa_size % 4 == 0);
+
+	end = fa->fa_size;
+	for (addr = 0; addr < end; addr += sizeof(data)) {
+		if (end - addr < sizeof(data)) {
+			bytes_to_read = end - addr;
+		} else {
+			bytes_to_read = sizeof(data);
+		}
+
+		rc = flash_area_read(fa, addr, data, bytes_to_read);
+		if (rc != 0) {
+			flash_area_close(fa);
+			return rc;
+		}
+
+		for (i = 0; i < bytes_to_read / 4; i++) {
+			if (data[i] != 0xffffffff) {
```

Comment:
max size of?

Reply:
It has rather a meaning of erase flash, which is usually 0xffffffff. This is however not true all the time, which it seems they have addressed in mcumgr code already (by reading flash property describing what is the value after flash erase operation).

We can either update to newest implementation from mcumgr, or drop erased flash checking (which is an optimization).

Reply:
What sort of effort would updating MCU manager be?
I assume we'd do that in a separate PR? which would merge first?

Reply:
Added new commit, that uses improved implementation of this function, from mcumgr.